### ISSUE TITLE
Porting some Layers from Gym

### ISSUE THREAD

**Issue by Sentient07** [Lasagne/Lasagne#825]
There are some layers in OpenAI [Gym](https://github.com/openai/improved-gan/blob/master/mnist_svhn_cifar10/nn.py) that would be helpful in training GANs. They're all Lasagne layers and it would be great to have them here directly. 
Namely, 
1) WeightNormLayer
2) MinibatchLayer
3) log_sum_exp

**Comment by f0k** [comment]
1. Their implementation is a bit cumbersome. I'd implement this in `lasagne.init` instead. But yes, it'd be nice to include weight normalization in Lasagne.
2. From https://groups.google.com/d/msg/lasagne-users/iFbmxHbmB9Q/be-WWuexDAAJ: "Hmm, not sure if it warrants direct inclusion in Lasagne. It'd be nice to have a small and tidy GAN demo in Lasagne/Recipes to grab it from, but on the other hand, OpenAI already provides the Lasagne layer (https://github.com/openai/improved-gan/blob/master/mnist_svhn_cifar10/nn.py#L132-L170)."
3. That's not really needed, Theano knows about this:
```python
>>> import theano
>>> import theano.tensor as T
>>> x = T.matrix('x')
>>> y = T.log(T.sum(T.exp(x), axis=1))
>>> theano.pp(theano.function([x], y).maker.fgraph.outputs[0])
'Elemwise{Composite{(i0 + log(i1))}}[(0, 0)](Reduce{maximum}{1}(x), Sum{axis=[1], acc_dtype=float64}(Elemwise{Composite{exp((i0 - i1))}}(x, DimShuffle{0, x}(Reduce{maximum}{1}(x)))))'
```

**Comment by Sentient07** [comment]
So there are two tasks in this. One is to implement GANs in recipes, the other is to implement the `WeightNormLayer`  right ? 
For including in Recipes, aren't the models that you have as gist good examples to be added there? I have few GAN implementations as ipython notebook,s which might be an example better visualization (By visualization I mean showing the image that our model generate. But probably that might be redundant as they can be seen in the original paper that I implemented as well). Also, I train them the way you have trained in your gist, so I think it would be better (in terms of originality) if you add it there :)
Regarding addition of WeightNorm layer in `lasagne.init`, I will open a PR after my current one is good to be merged. 
Thanks,

**Comment by Sentient07** [comment]
A small confusion regarding inclusion in recipes, should be done like how it is done in the paper "Improved Technique for training GANs" ?

**Comment by f0k** [comment]
> A small confusion regarding inclusion in recipes, should the implementation be done like how it is done in the paper "Improved Technique for training GANs" ?

If the idea was to have a reference for the minibatch discrimination layer, then yes. My gists only show how to train GANs based on the DCGAN paper, they don't include fancier models.

> Regarding addition of WeightNorm layer in `lasagne.init`

Note that it should not become a layer, but an initializer that you can wrap other initializers in, so you can do `l = DenseLayer(l, 20, W=lasagne.init.WeightNorm(lasagne.init.GlorotUniform()))`. The mean-only batch normalization they use would be separate from that.

**Comment by Sentient07** [comment]
> If the idea was to have a reference for the minibatch discrimination layer, then yes. My gists only show how to train GANs based on the DCGAN paper, they don't include fancier models.

Isn't it just changing the models of generation and discrimination to fit the paper? Or, does the way they train vary apart from the "Improved Technique for training GANs" paper ? Also, which other GANs would like to be added ?

**Comment by f0k** [comment]
> Isn't it just changing the models of generation and discrimination to fit the paper?

Yes, probably. I was referring to your question "For including in Recipes, aren't the models that you have as gist, good examples to be added there?", and my answer was that they don't include a minibatch discrimination layer, so they're not the kind of "small and tidy GAN demo in Lasagne/Recipes to grab it from" I meant in https://github.com/Lasagne/Lasagne/issues/825#issuecomment-292912784.

**Comment by Sentient07** [comment]
How to test the Weight normalization ? Is it fine if i have a numpy implementation, normalizing rows and columns as two different functions and compare with what the layer returns? (similar to `batch_norm` )
